java/core/src/main/java/com/microsoft/bond/Bonded.java

```diff
@@ -3,28 +3,117 @@
 
 package com.microsoft.bond;
 
+import com.microsoft.bond.helpers.ArgumentHelper;
 import com.microsoft.bond.protocol.ProtocolWriter;
 
 import java.io.IOException;
+import java.io.InputStream;
 
 /**
- * Placeholder Bonded for initial development. Just a box around a T.
+ * A construct representing a Bonded payload of some struct type. If the struct type is known in advance,
+ * a Bonded instance may be constrained to that type by specifying the type parameter; otherwise, the
+ * type parameter may be left unconstrained (i.e. the constraint is just the {@link BondSerializable}).
+ *
+ * @param <T> constraint on the payload type.
  */
-public class Bonded<T extends BondSerializable> implements IBonded<T> {
-    private T value;
+public abstract class Bonded<T extends BondSerializable> {
 
-    public Bonded(T instance) {
-        value = instance;
+    private final StructBondType<T> bondType;
+
+    /**
+     * Package-private constructor to restrict inheritance to the Bond Core package only.
+     *
+     * @param bondType the type descriptor of the struct type or null if the type is unknown
+     */
+    Bonded(StructBondType<T> bondType) {
+        this.bondType = bondType;
+    }
+
+    /**
+     * Returns the type descriptor of the struct type of this Bonded instance or null if the type if unspecified.
+     *
+     * @return type descriptor or null if unspecified.
+     */
+    public final StructBondType<T> getBondType() {
+        return this.bondType;
     }
 
-    @Override
-    public void serialize(ProtocolWriter writer) throws IOException {
-        // TODO: implement serialization/deserialization of Bonded types
+    /**
+     * Creates a new Bonded instance from an input stream that is constrained to a given struct type.
+     *
+     * @param inputStream input stream containing the serialized representation of this Bonded value
+     * @param bondType    the type descriptor of the struct type
+     * @param <T>         the struct type
+     * @return new Bonded instance backed by the stream
+     * @throws IllegalArgumentException if an argument is null
+     */
+    public static <T extends BondSerializable> Bonded<T> fromStream(
```

Comment:
As a human reader, the `T` type parameters on these static methods shadow the class-level `T`. I know that the Java compiler doesn't seem them as shadowing, but consider using a different parameter name (`U`?) to make it easier for the humans to read.